kedro/config/common.py

```diff
@@ -154,6 +158,11 @@ def _load_config_file(config_file: Path, ac_template: bool = False) -> Dict[str,
     except AttributeError as exc:
         raise BadConfigException(f"Couldn't load config file: {config_file}") from exc
 
+    except ParserError as exc:
+        traceback.print_exc()
```

Comment:
How does it look different using `traceback` from raising the error directly?

Reply:
Agreed it would be nice to see a screenshot!

Reply:
When raising the error directly the warning has to be printed before the error is raised, and the message gets lost at the top of the traceback. This workaround allows it so that the additional warning is the most recent output to console.

Reply:
@AhdraMeraliQB I see, would it be sufficient to use the same pattern as follow so the custom warning show at the bottom?:

`raise ParserError(f"Failed to parse file: {config_file} is invalid.") from exec`?

![image](https://user-images.githubusercontent.com/18221871/162012870-688d73c7-2945-4ce9-8938-1d892c688052.png)

Reply:
@noklam Whilst this solution is more elegant in code, I am concerned that the pointer to where the error occurred will be buried:

```
Traceback (most recent call last):
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/kedro/config/common.py", line 155, in _load_config_file
    for k, v in anyconfig.load(yml, ac_template=ac_template).items()
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/anyconfig/api.py", line 484, in load
    **options)
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/anyconfig/api.py", line 358, in single_load
    ac_context=ac_context, **options)
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/anyconfig/api.py", line 301, in _single_load
    return psr.load(ioi, **options)
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/anyconfig/backend/base.py", line 259, in load
    cnf = self.load_from_stream(ioi.src, container, **options)
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/anyconfig/backend/base.py", line 54, in wrapper
    return func(*args[1:], **kwargs)
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/anyconfig/backend/yaml/pyyaml.py", line 174, in yml_load
    **common.filter_from_options("ac_dict", options))
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/anyconfig/backend/yaml/pyyaml.py", line 151, in yml_fnc
    return fnc(*args, **common.filter_from_options("ac_safe", options))
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/yaml/__init__.py", line 114, in load
    return loader.get_single_data()
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/yaml/constructor.py", line 49, in get_single_data
    node = self.get_single_node()
  File "yaml/_yaml.pyx", line 707, in yaml._yaml.CParser.get_single_node
  File "yaml/_yaml.pyx", line 725, in yaml._yaml.CParser._compose_document
  File "yaml/_yaml.pyx", line 776, in yaml._yaml.CParser._compose_node
  File "yaml/_yaml.pyx", line 892, in yaml._yaml.CParser._compose_mapping_node
  File "yaml/_yaml.pyx", line 905, in yaml._yaml.CParser._parse_next_event
yaml.parser.ParserError: while parsing a block mapping
  in "/Users/ahdra_merali/Desktop/kedrotest/conf/base/catalog.yml", line 46, column 1
did not find expected key
  in "/Users/ahdra_merali/Desktop/kedrotest/conf/base/catalog.yml", line 48, column 3
```
**The original error message with file, line and token**
```
The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/anaconda3/envs/kedrotest/bin/kedro", line 8, in <module>
    sys.exit(main())
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/kedro/framework/cli/cli.py", line 206, in main
    cli_collection()
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/kedro/framework/cli/cli.py", line 146, in main
    **extra,
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/kedro/framework/cli/project.py", line 361, in run
    pipeline_name=pipeline,
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/kedro/framework/session/session.py", line 394, in run
    load_versions=load_versions,
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/kedro/framework/context/context.py", line 267, in _get_catalog
    conf_catalog = self._config_loader.get("catalog*", "catalog*/**", "**/catalog*")
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/kedro/config/config.py", line 100, in get
    conf_paths=self.conf_paths, patterns=list(patterns)
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/kedro/config/common.py", line 108, in _get_config_from_patterns
    config_filepaths=config_filepaths, ac_template=ac_template
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/kedro/config/common.py", line 192, in _load_configs
    single_config = _load_config_file(config_filepath, ac_template=ac_template)
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/kedro/config/common.py", line 165, in _load_config_file
    raise ParserError(f"Failed to parse file: {config_file} is invalid.") from exc
yaml.parser.ParserError: Failed to parse file: /Users/ahdra_merali/Desktop/kedrotest/conf/base/catalog.yml is invalid.

```
**The custom warning to explain the error**

Reply:
This fragment `"/Users/ahdra_merali/Desktop/kedrotest/conf/base/catalog.yml", line 48, column 3` is by far the most useful to users, obviously don't go too far down the rabbit hole on this one - but it would amazing if we could extract it somehow and present to the user.

Reply:
@datajoely The current implementation prints a traceback that looks like this:

```
Traceback (most recent call last):
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/kedro/config/common.py", line 155, in _load_config_file
    for k, v in anyconfig.load(yml, ac_template=ac_template).items()
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/anyconfig/api.py", line 484, in load
    **options)
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/anyconfig/api.py", line 358, in single_load
    ac_context=ac_context, **options)
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/anyconfig/api.py", line 301, in _single_load
    return psr.load(ioi, **options)
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/anyconfig/backend/base.py", line 259, in load
    cnf = self.load_from_stream(ioi.src, container, **options)
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/anyconfig/backend/base.py", line 54, in wrapper
    return func(*args[1:], **kwargs)
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/anyconfig/backend/yaml/pyyaml.py", line 174, in yml_load
    **common.filter_from_options("ac_dict", options))
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/anyconfig/backend/yaml/pyyaml.py", line 151, in yml_fnc
    return fnc(*args, **common.filter_from_options("ac_safe", options))
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/yaml/__init__.py", line 114, in load
    return loader.get_single_data()
  File "/usr/local/anaconda3/envs/kedrotest/lib/python3.7/site-packages/yaml/constructor.py", line 49, in get_single_data
    node = self.get_single_node()
  File "yaml/_yaml.pyx", line 707, in yaml._yaml.CParser.get_single_node
  File "yaml/_yaml.pyx", line 725, in yaml._yaml.CParser._compose_document
  File "yaml/_yaml.pyx", line 776, in yaml._yaml.CParser._compose_node
  File "yaml/_yaml.pyx", line 892, in yaml._yaml.CParser._compose_mapping_node
  File "yaml/_yaml.pyx", line 905, in yaml._yaml.CParser._parse_next_event
yaml.parser.ParserError: while parsing a block mapping
  in "/Users/ahdra_merali/Desktop/kedrotest/conf/base/catalog.yml", line 46, column 1
did not find expected key
  in "/Users/ahdra_merali/Desktop/kedrotest/conf/base/catalog.yml", line 48, column 3
Failed to parse file: /Users/ahdra_merali/Desktop/kedrotest/conf/base/catalog.yml is invalid.
```

With that fragment immediately followed by the warning - is this sufficient or would you suggest separating this out further?

Reply:
I actually think the relevant attributes are within the Exception object - the class inherits from something called [MarkedYamlErorr](https://github.com/yaml/pyyaml/blob/8cdff2c80573b8be8e8ad28929264a913a63aa33/lib/yaml/error.py#L48) and there is an attribute called `problem_mark` which has the line and other information